src/frontend/src/optimizer/plan_node/logical_join.rs

```diff
@@ -702,26 +788,94 @@ impl PredicatePushdown for LogicalJoin {
             LogicalJoin::can_push_on_from_filter(join_type),
         );
 
-        let mut new_on = self.on.clone().and(on);
-        let (left_from_on, right_from_on, on) = LogicalJoin::push_down(
-            &mut new_on,
+        let mut on = self.on.clone().and(on);
+        let (left_from_on, right_from_on, old_on) = LogicalJoin::push_down(
+            &mut on,
             left_col_num,
             right_col_num,
             LogicalJoin::can_push_left_from_on(join_type),
             LogicalJoin::can_push_right_from_on(join_type),
             false,
         );
         assert!(
-            on.always_true(),
+            old_on.always_true(),
             "On-clause should not be pushed to on-clause."
         );
 
         let left_predicate = left_from_filter.and(left_from_on);
         let right_predicate = right_from_filter.and(right_from_on);
 
-        let new_left = self.left.predicate_pushdown(left_predicate);
-        let new_right = self.right.predicate_pushdown(right_predicate);
-        let new_join = LogicalJoin::new(new_left, new_right, join_type, new_on);
+        let mut left = self.left.predicate_pushdown(left_predicate);
+        let mut right = self.right.predicate_pushdown(right_predicate);
+        let mut output_indices = self.output_indices().to_owned();
+        // This code block pushes the calculation of inputs of join's comparison condition to join's
+        // two sides.
+        {
+            let exprs = on.conjunctions;
+            let (left_exprs, right_exprs, indices_and_ty_of_func_calls) =
+                Self::find_comparison_exprs(left_col_num, right_col_num, &exprs);
+
+            // used to shift indices of input_refs pointing the right side of `join` with
+            // `left_exprs.len`.
+            let mut col_index_mapping = {
+                let map = (0..left_col_num)
+                    .chain(
+                        (left_col_num..left_col_num + right_col_num).map(|i| i + left_exprs.len()),
+                    )
+                    .map(Some)
+                    .collect_vec();
+                ColIndexMapping::new(map)
+            };
+            let (mut exprs, new_output_indices) =
+                Self::remap_exprs_and_output_indices(exprs, output_indices, &mut col_index_mapping);
+            output_indices = new_output_indices;
+
+            // ```ignore
+            // the internal table of join has has the following schema:
+            // original left's columns | left_exprs | original right's columns | right_exprs
+            //```
+            // `left_index` and `right_index` will scan through `left_exprs` and `right_exprs`
+            // respectively.
+            let mut left_index = left_col_num;
+            let mut right_index = left_col_num + left_exprs.len() + right_col_num;
+            // replace chosen function calls.
+            for (i, (index_of_func_call, ty)) in
+                indices_and_ty_of_func_calls.into_iter().enumerate()
```

Comment:
```
for ((left_expr, right_expr), (index_of_func_call, ty)) in left_exprs
    .iter()
    .zip_eq(&right_exprs)
    .zip_eq(indices_and_ty_of_func_calls)
```
to avoid indexing `left_exprs[i]`